instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt

```diff
@@ -0,0 +1,547 @@
+        )
+    }
+
+    @Test
+    fun routeRefreshIsNotCancelledOnDestroyTest() = sdkTest {
+        setupMockRequestHandlers(
+            coordinates,
+            R.raw.route_response_single_route_refresh,
+            NthAttemptHandler(
+                createRefreshHandler(
+                    R.raw.route_response_route_refresh_annotations,
+                    "route_response_single_route_refresh"
+                ),
+                2
+            )
+        )
+
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(5_000))
+        mapboxNavigation.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.startTripSession()
+        val requestedRoutes = requestRoutes()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(requestedRoutes)
+        delay(8000) // refresh interval + accuracy
+
+        mapboxNavigation.onDestroy()
+        delay(2000) // accuracy: to wait for potential cancelled
+
+        assertEquals(
+            listOf(RouteRefreshExtra.REFRESH_STATE_STARTED),
+            observer.getStatesSnapshot()
+        )
+    }
+
+    @Test
+    fun successfulRouteRefreshOnDemandTest() = sdkTest {
+        setupMockRequestHandlers(
+            coordinates,
+            R.raw.route_response_single_route_refresh,
+            createRefreshHandler(
+                R.raw.route_response_route_refresh_annotations,
+                "route_response_single_route_refresh"
+            )
+        )
+
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(10_000))
+        mapboxNavigation.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.startTripSession()
+        val requestedRoutes = requestRoutes()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(requestedRoutes)
+        mapboxNavigation.refreshRoutesImmediately()
+
+        waitForRefresh()
+
+        assertEquals(
+            listOf(
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_FINISHED_SUCCESS
+            ),
+            observer.getStatesSnapshot()
+        )
+    }
+
+    @Test
+    fun failedRouteRefreshOnDemandTest() = sdkTest {
+        setupMockRequestHandlers(
+            coordinates,
+            R.raw.route_response_single_route_refresh,
+            NthAttemptHandler(
+                createRefreshHandler(
+                    R.raw.route_response_route_refresh_annotations,
+                    "route_response_single_route_refresh"
+                ),
+                1
+            )
+        )
+
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(10_000))
+        mapboxNavigation.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.startTripSession()
+        val requestedRoutes = requestRoutes()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(requestedRoutes)
+        mapboxNavigation.refreshRoutesImmediately()
+
+        waitForRefresh()
```

Comment:
It's the previous behaviour: if routes were changed after 3 failed refresh attempts (and they are changed because we remove expiring data from them), we notify the user. Again, I don't think this behaviour should be broken.

Reply:
Where do the 3 failed request come from in this test though? We're making one on-demand request and expect the failure, why are we waiting for the 3 scheduled requests?

3 request with interval of 10 seconds (as is the test configuration) should exceed our test timeout anyway, so I still don't understand where is this coming from.

Reply:
Ah, I didn't understand the question.

> Why does a failed refresh attempt still invoke RoutesObserver with reason `RoutesExtra.ROUTES_UPDATE_REASON_REFRESH`?

I think it should be like this for on demand refresh. You trigger the request - you get the result. Either success or failure. If it's a failure, we remove expiring data from route -> we get new refreshed routes.

Reply:
> If it's a failure, we remove expiring data from route -> we get new refreshed routes.

I think this is debatable. We're risking clearing all the data on a single failed request whereas the regular cycle requires all 3 requests and their intervals. A small gap in network connectivity (like a request made in a tunnel) would lead to pretty poor user experience.

Alternative would be to only clear the data on 3 unconditional requests but since we reset the counter, if someone calls the on-demand refresh often, we'd never clear the data. This is not ideal but still sounds better to me - clearing the data is a last resort when we're sure the data is already stale, this is not the case after a single failed on-demand refresh.

Reply:
Agreed that it should not be like this. But I think it would be better to use time since last refresh, not the number of attempts. Implemented it in a separate commit. BTW, it also solves the issue with invoking `removeExpiredDataFromRoute` unnecessarily often.